src/platform/w3dengine/client/w3dingameui.cpp

```diff
@@ -13,9 +13,225 @@
  *            LICENSE
  */
 #include "w3dingameui.h"
+#include "gameclient.h"
+#include "gamewindowmanager.h"
+#include "globaldata.h"
+#include "hanim.h"
+#include "rendobj.h"
+#include "terrainlogic.h"
+#include "w3dassetmanager.h"
+#include "w3ddisplay.h"
+#include "w3dscene.h"
 #include "w3dview.h"
 
+W3DInGameUI::W3DInGameUI() : m_locator01(nullptr), m_locator02(nullptr)
+{
+    for (int i = 0; i < MAX_MOVE_HINTS; i++) {
+        m_moveHintRenderObj[i] = nullptr;
+        m_moveHintAnim[i] = nullptr;
+    }
+}
+
+W3DInGameUI::~W3DInGameUI()
+{
+    for (int i = 0; i < MAX_MOVE_HINTS; i++) {
+        Ref_Ptr_Release(m_moveHintRenderObj[i]);
+        Ref_Ptr_Release(m_moveHintAnim[i]);
+    }
+
+    Ref_Ptr_Release(m_locator01);
+    Ref_Ptr_Release(m_locator02);
+}
+
+void W3DInGameUI::Init()
+{
+    InGameUI::Init();
+}
+
+void W3DInGameUI::Reset()
+{
+    InGameUI::Reset();
+}
+
+void W3DInGameUI::Update()
+{
+    InGameUI::Update();
+}
+
+void W3DInGameUI::Draw()
+{
+    Pre_Draw();
+
+    if (m_isDragSelecting) {
+        Draw_Selection_Region();
+    }
+
+    if (g_theDisplay != nullptr) {
+        for (View *view = g_theDisplay->Get_First_View(); view != nullptr; view = g_theDisplay->Get_Next_View(view)) {
+            Draw_Move_Hints(view);
+            Draw_Attack_Hints(view);
+            Draw_Place_Angle(view);
+        }
+    }
+
+    Post_Draw();
+    g_theWindowManager->Win_Repaint();
+}
+
 View *W3DInGameUI::Create_View()
 {
     return new W3DView();
 }
+
+void W3DInGameUI::Draw_Selection_Region()
+{
+    g_theDisplay->Draw_Open_Rect(m_dragSelectRegion.lo.x,
+        m_dragSelectRegion.lo.y,
+        m_dragSelectRegion.hi.x - m_dragSelectRegion.lo.x,
+        m_dragSelectRegion.hi.y - m_dragSelectRegion.lo.y,
+        2.0f,
+        0x9933FF33);
+}
+
+void W3DInGameUI::Draw_Move_Hints(View *view)
+{
+    int count = 0;
+    Matrix3D tm;
+
+    for (;;) {
+        if ((g_theGameClient->Get_Frame() - m_moveHint[count].frame) > 40) {
+            if (m_moveHintRenderObj[count] != nullptr && !m_moveHintRenderObj[count]->Is_Hidden()) {
+                m_moveHintRenderObj[count]->Set_Hidden(true);
+                W3DDisplay::s_3DScene->Remove_Render_Object(m_moveHintRenderObj[count]);
+            }
+
+            goto label1;
+        }
+
+        if (m_moveHintRenderObj[count] == nullptr) {
+            break;
+        }
+
+    label2:
```

Comment:
Is there no version of this function without this label?

Reply:
I can't find anything cleaner.